src/wpool/gun_server.erl

```diff
@@ -0,0 +1,92 @@
+
+-include("mongoose_logger.hrl").
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+start_link(Destination, Options) ->
+    gen_server:start_link(?MODULE, {Destination, Options}, []).
+
+%%%===================================================================
+%%% gen_server callbacks
+%%%===================================================================
+
+-spec init({{inet:hostname() | inet:ip_address(), inet:port_number()}, gun:opts()}) ->
+    {ok, State :: #state{}}.
+init({{Host, Port}, Opts}) ->
+    {ok, #state{host = Host, port = Port, opts = Opts}}.
+
+-spec handle_call(Request :: request(), From :: {pid(), Tag :: term()}, State :: #state{}) ->
+                     {reply, Reply :: term(), NewState :: #state{}}.
+handle_call({request, FullPath, Method, Headers, Query, _Retries, Timeout}, _From, State) ->
+    LHeaders = lowercase_headers(Headers),
+
+    {ok, PID} = gun:open(State#state.host, State#state.port, maps:merge(default_opts(), State#state.opts)),
+    {ok, _Protocol} = gun:await_up(PID),
+
+    StreamRef = gun:request(PID, Method, FullPath, LHeaders, Query),
+    case gun:await(PID, StreamRef, Timeout) of
+        {response, fin, Status, _H} ->
+            {reply, {ok, {{integer_to_binary(Status), reason}, resp_headers, no_data, size, time}}, State};
+        {response, nofin, Status, _H} ->
+            {ok, B} = gun:await_body(PID, StreamRef),
+            {reply, {ok, {{integer_to_binary(Status), reason}, resp_headers, B, size, time}}, State};
+        {error, timeout} ->
+            {reply, {error, request_timeout}, State};
+        {E, R} ->
+            ?WARNING_MSG("Error in gun_server: ~p, Reason: ~p", [E, R]),
+            {reply, {E, R}, State}
+    end.
+
+-spec handle_cast(Request :: request(), State :: #state{}) -> {noreply, NewState :: #state{}}.
+handle_cast({request, FullPath, Method, Headers, Query, _Retries, Timeout}, State) ->
+    LHeaders = lowercase_headers(Headers),
+
+    {ok, PID} = gun:open(State#state.host, State#state.port, maps:merge(default_opts(), State#state.opts)),
+    {ok, _Protocol} = gun:await_up(PID),
+
+    StreamRef = gun:request(PID, Method, FullPath, LHeaders, Query),
+    case gun:await(PID, StreamRef, Timeout) of
+        {response, fin, _Status, _H} ->
+            {noreply, State};
+        {response, nofin, _Status, _H} ->
+            {ok, _B} = gun:await_body(PID, StreamRef),
+            {noreply, State};
+        {E, R} ->
+            ?WARNING_MSG("Error in gun_server: ~p, Reason: ~p", [E, R]),
+            {noreply, State}
+    end.
+
+%%%===================================================================
+%%% Internal functions
+%%%===================================================================
+
+default_opts() ->
+    #{protocols => [http2]}.
+
+lowercase_headers(Headers) ->
+    [{string:lowercase(K), V} || {K, V} <- Headers].
```

Comment:
End-of-line missing, unix traditionally doesn't like that (note the 🚫 symbol here in github). If you're using VSC I recommend this one: https://github.com/riccardoNovaglia/vsCodeBlankLine